### DISCUSSION TITLE
Production readiness of object storge support in Merge Tree

### DISCUSSION THREAD

**Discussion by goelrajat** [ClickHouse/ClickHouse#38216] [Q&A]
We want to use Merge Tree Storage Tiering support on s3 disk (using TTL moves) for our production deployments, in next couple of months. The documentation page (https://clickhouse.com/docs/en/integrations/s3/s3-merge-tree) says the feature is experimental. Our tables are replicated but we can live without zero copy replication or data caching optimizations. Can we use this feature in production ? If not, what is the plan for its production readiness?

**Comment by UnamedRus** [comment]
> S3 Disk

Yes, it's production ready.
Some of our clients use it for more than 1 year.

> Zero copy replication

Zero copy replication has some issues, but in GA for yandex.cloud.

**Reply by alexey-milovidov** [reply]
Zero-copy replication is not production-ready.
There are many known bugs, including data corruption.
Please disregard the answer from @UnamedRus.

**Reply by Anton-Rodionov** [reply]
@alexey-milovidov Given the GA of Clickhouse Cloud, is zero-copy replication production-ready now?

<img width="1127" alt="Screenshot 2022-12-08 at 00 24 59" src="https://user-images.githubusercontent.com/5286337/206318945-3ccf8af6-128e-4529-bf80-ebbe9a838376.png">

**Reply by alexey-milovidov** [reply]
@Anton-Rodionov, we discovered and fixed a bug in zero-copy replication and projections just a few days ago:
https://github.com/ClickHouse/ClickHouse/pull/43764

I'd say it's still not recommended for production usage. It's ok to use under supervision.

**Reply by Anton-Rodionov** [reply]
@alexey-milovidov, thank you, we have tight control over the queries and can give it a try.

What would be your suggestion about migrating towards zero-copy replication model? In a "classical" model of bi-level sharding (N layers x M shards, where data is sharded first by e.g. company_id and then by user_id) one can either follow the Clickhouse Cloud model where all hosts are replicas of each other and "look at" the same S3 prefix (no sharding at all); alternatively one can leave layers/shards and have a separate S3 prefix for replicas within each shard.

The benefits of the Clickhouse Cloud approach is great flexibility and simplicity, but I'm a little concerned about data locality. The parts will become larger (or will be too many of them) as a single part now contains data from all shards, page cache might not be as efficient, and if we stop sharding by user_id the queries like uniq(user_id) will have to merge partial aggregates instead of relying on distributed_group_by_no_merge.

I'll of course experiment, but I'm sure you have put a lot of thoughts into this.

**Reply by alexey-milovidov** [reply]
Yes, data locality is a concern. It can work similarly to a configuration with a sharding key if a query will request the reading of filtered data on every replica (filtered by the remainder of the division by the sharding key). Then if the primary key starts with the sharding key, the data locality will be the same as for sharded cluster.

**Reply by Anton-Rodionov** [reply]
This makes sense, thank you!

**Comment by alexey-milovidov** [comment]
This (benign) error is also a bug in the implementation of "zero-copy" replication:
https://github.com/ClickHouse/ClickHouse/issues/43185

**Reply by Anton-Rodionov** [reply]
Does Clickhouse Cloud use zero-copy replication? If it's under NDA please disregard the question.
